### PR TITLE
i#1312 AVX-512 support: Add instructions vmovlpd, vmovsldup, vmovhps, vmovhpd, vmovshdup.

### PR DESCRIPTION
**Pull request by hgreving2304**
Adds the evex form of instructions vmovsldup, vmovddup, vmovhps, vmovhpd, vmovshdup.

Adds the new subreg type OPSZ_half_16_vex32_evex64 for AVX-512.

Adds the decoder flag and check for EVEX.LL.

Adds tests for instructions above.

Issue: #1312

### CODE REVIEW

File: core/arch/decode.h [hunk 1]
```diff
@@ -393,9 +393,12 @@ enum {
     OPSZ_half_16_vex32,      /* half of 128 bits (XMM or memory);
                               * if vex.L then is half of 256 bits (YMM or memory).
                               */
-    /* XXX i#1312: Augment with new types specific to AVX-512. */
+    OPSZ_half_16_vex32_evex64, /* 64 bits, but can be half of XMM register;
+                                * if evex.L then is 256 bits (YMM or memory);
+                                * if evex.L' then is 512 bits (ZMM or memory)
```
Comment (derekbruening) [inline core/arch/decode.h]: nit: missing trailing `.`

File: core/arch/x86/decode.c [hunk 2]
```diff
@@ -1179,9 +1186,11 @@ read_instruction(byte *pc, byte *orig_pc, const instr_info_t **ret_info,
     } else if (info != NULL && di->evex_encoded) {
         if (!TEST(REQUIRES_EVEX, info->flags))
             info = NULL; /* invalid encoding */
-        /* XXX i#1312: This might need checks for REQUIRES_EVEX_L_0 and
-         * REQUIRES_EVEX_LL_0.
-         */
+        else if (TEST(REQUIRES_VEX_L_0, info->flags) && TEST(PREFIX_VEX_L, di->prefixes))
+            info = NULL;
```
Comment (derekbruening) [inline core/arch/x86/decode.c]: nit: all 3 of the others above have a comment; this looks like it's somehow different w/o it, but it's not, right?  More readable when consistent.

File: core/arch/x86/decode_table.c [hunk 3]
```diff
@@ -1498,6 +1498,7 @@ const instr_info_t * const op_instr[] =
 #define Wed TYPE_W, OPSZ_16_vex32_evex64
 #define Vex TYPE_V, OPSZ_16_vex32_evex64
 #define Wex TYPE_W, OPSZ_16_vex32_evex64
+#define Weh_x TYPE_W, OPSZ_half_16_vex32_evex64
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: wrong code: should be `Wh_e`

File: core/arch/x86/encode.c [hunk 4]
```diff
@@ -868,24 +882,29 @@ reg_size_ok(decode_info_t *di /*prefixes field is IN/OUT; x86_mode is IN*/, reg_
         opnd_size_t expanded = expand_subreg_size(opsize);
         if (expanded == OPSZ_8 &&
             (optype == TYPE_P || optype == TYPE_Q || optype == TYPE_P_MODRM))
-            return (reg >= REG_START_MMX && reg <= REG_STOP_MMX);
+            return (reg >= DR_REG_START_MMX && reg <= DR_REG_STOP_MMX);
```
Comment (derekbruening) [inline core/arch/x86/encode.c]: These seem like an unrelated cleanup for a different PR.

File: core/arch/x86/decode_table.c [hunk 5]
```diff
@@ -1498,6 +1498,7 @@ const instr_info_t * const op_instr[] =
 #define Wed TYPE_W, OPSZ_16_vex32_evex64
 #define Vex TYPE_V, OPSZ_16_vex32_evex64
 #define Wex TYPE_W, OPSZ_16_vex32_evex64
```
Comment (derekbruening) [inline core/arch/x86/decode_table.c]: These all look wrong too: 'e' means 'dq, qq, or oq', i.e., OPSZ_16_vex32_evex64.  Why is there another letter after the 'e' on these?

File: core/arch/decode.h (hunk 1 above)
Comment (hgreving2304) [reply core/arch/decode.h]: Done.

File: core/arch/x86/decode.c (hunk 2 above)
Comment (hgreving2304) [reply core/arch/x86/decode.c]: Done.

File: core/arch/x86/decode_table.c (hunk 3 above)
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: So we call '_x' either 16, vex32 and _e 16, vex32, or evex64? To me the '_x' meant 'any vector' while the 'e' addition to the type's size meant evex. We had discussed this and came up with Wex etc. It's all subject to interpretation that's why I wouldn't call it "wrong". I personally actually don't care whether it is Weh_x or Wh_e. So let's use '_e' for as elaborated here, but there may be other corner case names coming up later.

Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Renamed.

File: core/arch/x86/encode.c (hunk 4 above)
Comment (hgreving2304) [reply core/arch/x86/encode.c]: Reverted.

File: core/arch/x86/decode_table.c (hunk 3 above)
Comment (derekbruening) [reply core/arch/x86/decode_table.c]: ?  We had already agreed on what these mean.  `n_m` means "n out of m".  `x` is Intel's code, and it means "dq or qq".  `e` is our new size code for "dq or qq or oq".

Comment (derekbruening) [reply core/arch/x86/decode_table.c]: There are also two-letter codes following Intel's {p,s,v}{s,d,i} that indicate the element type of a vector.  We decided to extend the {p,s,v} to add an 'e' to mean, again, 128, 256, or 512.  So e{s,d,i} is a valid code for a packed vector, though not all that useful to DR since the IR doesn't track the element type.

File: core/arch/x86/decode_table.c (hunk 5 above)
Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: We came up with this offline during another review. https://github.com/DynamoRIO/dynamorio/compare/91d1d006ed2f...58da5203eb41
AFAI can tell Intel uses Vs, Vd, Vx for type 'V' and single precision, double precision and 'x' for don't care. We enhanced it with 'e' for the invented evex version of 'V'. If you want Ve, how do you distinguish between Vs, Vd and Vx then for evex?

Comment (derekbruening) [reply core/arch/x86/decode_table.c]: I have never seen `Vs` or `Vd`.  Do you mean `vs` rather than `Vs`?  They are quite different.  But then there is no such things as `vx`.  See my comment above about {p,s,v,e}{s,d,i}.

Comment (derekbruening) [reply core/arch/x86/decode_table.c]: > We came up with this offline during another review

Yes the `ed` and `es` make sense for double and single float in a size `e` vector.  But the `ex` should just be `x`, just like the `dq` and other full-size-only codes that don't seem to care what the component types are.

Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Whatever the name is, we still need to distinguish TYPE_W, OPSZ_16_vex32_evex64 and TYPE_W, OPSZ_16_vex32. I can name the latter Wx, and the former We?

Comment (derekbruening) [reply core/arch/x86/decode_table.c]: Yes, that is what they would be, We and Wx.

Comment (hgreving2304) [reply core/arch/x86/decode_table.c]: Done.
